_posts/2019-12-17-zero-to-main-rust-1.md

```diff
@@ -0,0 +1,433 @@
+        // These symbols come from `linker.ld`
+        static mut __sbss: u32; // Start of .bss section
+        static mut __ebss: u32; // End of .bss section
+        static mut __sdata: u32; // Start of .data section
+        static mut __edata: u32; // End of .data section
+        static __sidata: u32; // Start of .rodata section
+    }
+
+    // Initialize (Zero) BSS
+    unsafe {
+        let mut sbss: *mut u32 = &mut __sbss;
+        let ebss: *mut u32 = &mut __ebss;
+
+        while sbss < ebss {
+            write_volatile(sbss, zeroed());
+            sbss = sbss.offset(1);
+        }
+    }
+
+    // Initialize Data
+    unsafe {
+        let mut sdata: *mut u32 = &mut __sdata;
+        let edata: *mut u32 = &mut __edata;
+        let mut sidata: *const u32 = &__sidata;
+
+        while sdata < edata {
+            write_volatile(sdata, read(sidata));
+            sdata = sdata.offset(1);
+            sidata = sidata.offset(1);
+        }
+    }
+
+    // Call user's main function
+    main()
+}
+```
+
+Phew! That was a lot at once, especially if you aren't familiar with Rust! Let's break that down one chunk at a time to explain the concepts in a little more detail:
+
+### Defining a function in Rust
+
+```rust
+pub fn reset_handler() -> ! {
+```
+
+This defines a function that is public, named `reset_handler`, that takes no arguments `()`, and that never returns `-> !`.
+
+In Rust, functions normally either don't return a value like this:
+
+```rust
+/// Returns nothing
+fn foo() { /* ... */ }
+```
+
+Or do return a value like this:
+
+```rust
+/// Returns a 32-bit unsigned integer
+fn bar() -> u32 { /* ... */ }
+```
+
+The `!` type, called the "Never type", means that this function will never return, or diverges. Since our reset handler never will return (where would it go?) we can tell Rust this, which may allow it to make certain optimizations at compile time.
+
+### A little help from the linker
+
+```rust
+extern "C" {
+    // These symbols come from `linker.ld`
+    static mut __sbss: u32; // Start of .bss section
+    static mut __ebss: u32; // End of .bss section
+    static mut __sdata: u32; // Start of .data section
+    static mut __edata: u32; // End of .data section
+    static __sidata: u32; // Start of .rodata section
+}
+```
+
+This section defines a number of static symbols which will be provided by our linker, namely the start and end of the sections that are important for our reset handler to know about. These symbols are defined in an `extern "C"` scope, which means two things:
+
+1. They will be provided sometime later, by another piece of code, or in this case, the linker itself
+2. They will be defined using the "C" style of ABI and naming conventions, which means they are implicitly `#[no_mangle]`
+
+Some of these symbols are also declared as `mut`, or "mutable". By default in Rust, all variables are immutable, or read-only. To make a variable mutable in Rust, you must explicitly mark it as `mut`.
```

Comment:
Maybe: add a note about `const` in C and the parallel between it and `mut`.

Reply:
Good call!